### DISCUSSION TITLE
Question about the theory of free energy minimization

### DISCUSSION THREAD

**Discussion by biglinn** [SSCHAcode/python-sscha#369] [Q&A] [answered]
Dear developers,

   I have learned the theory of free energy minimization. I briefly summarize it as follows:
<img width="365" alt="屏幕截图 2024-12-16 164036" src="https://github.com/user-attachments/assets/e7843eca-eae8-4a9b-8ee3-3cd73b601a7d" />

I believe that the final _Phi_ is the dyn matrix output by `dyn.save_qe`.
Moreover, is the final _**R**_ saved?  Could it represent the equilibrium configuration at the specified temperature？

**Comment by mesonepigreco** [comment, marked as answer]
Hi, yes, the dyn.save_qe saves both $\Phi$ and $\mathcal R$, and, indeed, it represets the equilibrium configuration at the specified temperature.
To access the structure, you can load the dynamical matrices saved at the end of the minimization and export it as
```python
dyn = CC.Phonons.Phonons("my_dyn")
dyn.structure.save_scf("my_structure.scf")
```

The structure will be saved in the standard espresso file format. You can also save it in cif via
```
dyn.structure.save_generic_file("my_structure.cif")
```
As the "CIF" extension, all standard file formats supported by ASE should work out of the box.

**Reply by biglinn** [reply]
Thank you! I get it.

And I have a question about the density matrix $\rho$.
I find that SSCHA directly give its expression, like this:
<img width="437" alt="屏幕截图 2024-12-16 223438" src="https://github.com/user-attachments/assets/faae8a9d-0e08-4090-89d5-d52549922592" />
In this, the $\Psi$ is:
<img width="334" alt="屏幕截图 2024-12-16 223704" src="https://github.com/user-attachments/assets/4ff8351f-4f7e-4d3a-b09d-c5f3b64281a5" />

Could you recommend me an acticle with derivation of the expression?

**Reply by biglinn** [reply]
Hi, there is not function `save_generic_file` in the class `Structure.Structure`, but `read_generic_file` exits. Dose SSCHA not support cif format now?

**Comment by biglinn** [comment]
Hi, @mesonepigreco 

I have a question about performing DFT calculations for minimization with different DFT packages.
I have used QE and VASP to perform DFT calculations respectively.

Using QE:
<img width="334" alt="QE" src="https://github.com/user-attachments/assets/c681c3a8-93f7-4676-aadc-f46957b4235a" />
<img width="337" alt="QE-phonon" src="https://github.com/user-attachments/assets/eeea09e6-6aeb-47fe-b58b-6a76d614b6fc" />
The gradient converges by 3 populations, each population with 20 configurations.
Although the effective sample size of each population have reached the threshold controlled by kong-liu parameter, it doesn't matter.

Using VASP:
<img width="337" alt="VASP" src="https://github.com/user-attachments/assets/97d0f00d-075f-4b96-b376-838471165d73" />
<img width="337" alt="VASP-phonon" src="https://github.com/user-attachments/assets/defd7b9c-81f1-44b8-a61f-57a645eaf8b3" />
When I use VASP, the free energy and FC gradient show great discontinuity between two populations. The convergence is not reached with 4 populations. Moreover, its frequency evolution is completely different from that by QE.

Why does this happen with VASP?

**Reply by biglinn** [reply]
When using VASP, I just convert the unit from eV to Ry (eV/A to Ry/Bohr, eV/A^3 to Ry/Bohr^3).
But I find that energies of QE and VASP have large difference.
Qe is like this:
![image](https://github.com/user-attachments/assets/e68d7221-e32f-4753-a668-2815937800ff)
VASP is like this:
![image](https://github.com/user-attachments/assets/c893ee03-e603-4ed7-ab45-df479b768ddb)

Although the absolute value of energy is meaningless, is there something else to change?

**Reply by biglinn** [reply]
Oh my god! It's my fault!
I change the order of atoms when transform pw.in to POSCAR, so that the order of atomic forces in `forces_populationX_X.dat` don't correspond to the order of atomic displacement in `u_populationX_X.dat`

**Reply by Guodonglin-cqust** [reply]
> Oh my god! It's my fault! I change the order of atoms when transform pw.in to POSCAR, so that the order of atomic forces in `forces_populationX_X.dat` don't correspond to the order of atomic displacement in `u_populationX_X.dat`

The atomic positions generated by QE are not in the same order as those in the POSCAR, leading to incorrect forces and an increase in free energy rather than a decrease. Do you have any methods to resolve this issue? Thank you.

**Reply by biglinn** [reply]
I dont have a direct method to solve it. So I write scripts to record the order of atomic positions in the pw.in and transform back after extracting energy, forces from OUTCAR.

**Reply by Guodonglin-cqust** [reply]
when you use the script, after VASP calculations, dose the free energy decrease with the steps? Would you mind sharing the script you wrote? Thank you.
